lib/algo/SnapshotDMD.h

```diff
@@ -0,0 +1,131 @@
+#define included_SnapshotDMD_h
+
+#include "DMD.h"
+#include <vector>
+
+namespace CAROM {
+
+class Vector;
+class SnapshotDMD : public DMD
+{
+public:
+
+    /**
+     * @brief Constructor. Basic DMD with uniform time step size.
+     *        Inherited directly from base DMD class.
+     *
+     * @param[in] dim              The full-order state dimension.
+     * @param[in] dt               The dt between samples.
+     * @param[in] alt_output_basis Whether to use the alternative basis for
+     *                             output, i.e. phi = U^(+)*V*Omega^(-1)*X.
+     * @param[in] state_offset     The state offset.
+     */
+    SnapshotDMD(int dim, double dt, bool alt_output_basis = false,
+                Vector* state_offset = NULL) : DMD(dim,dt,alt_output_basis,state_offset) {}
+
+    /**
+     * @brief Constructor. DMD from saved models. Inherited directly
+     *        from base DMD class.
+     *
+     * @param[in] base_file_name The base part of the filename of the
+     *                           database to load when restarting from a save.
+     */
+    SnapshotDMD(std::string base_file_name) : DMD(base_file_name) {}
+
+    /**
+    * @brief Destroy the SnapshotDMD object
+    */
+    ~SnapshotDMD();
+
+    /**
+    * @brief Interpolate the current snapshots to n, new snapshots
+    *        distributed uniformly over the currently sampled time
+    *        domain.
+    * @param[in] n                  The number of desired snapshots.
+    */
+    void interpolateToNSnapshots(int n);
+
+    /**
+     * @brief Train the DMD model with specified reduced dimension. If k is
+     *        too large then new snapshots are computed using
+     *        interpolateToNSnapshots(k+1).
+     *
+     * @param[in] k               The number of modes to keep after doing SVD.
+     * @param[in] W0              The initial basis to prepend to W.
+     * @param[in] linearity_tol   The tolerance for determining whether a column
+     *                            of W is linearly independent with W0.
+     */
+    void train(int k, const Matrix* W0 = NULL, double linearity_tol = 0.0);
+
+    /**
+     * @brief Train the DMD model with energy fraction criterion.
+     *
+     * @param[in] energy_fraction The energy fraction to keep after doing SVD.
+     * @param[in] W0              The initial basis to prepend to W.
+     * @param[in] linearity_tol   The tolerance for determining whether a column
+     *                            of W is linearly independent with W0.
+     */
+    void train(double energy_fraction, const Matrix* W0 = NULL,
+               double linearity_tol = 0.0);
+
+    /**
+     * @brief Returns a copy of the current snapshot vector "d_snapshots"
+     */
+    std::vector<Vector*> getSnapshotVectors()
+    {
+        std::vector<Vector*> return_snapshots(d_snapshots);
+        return return_snapshots;
+    }
+protected:
+    /**
+     * @brief Obtain DMD model interpolant at desired parameter point by
```

Comment:
Since this is just declaring the existing function as a `friend`, I don't think we need to copy the comments here. They are already in ParametricDMD.h. Maybe just a 1-line comment like `See ParametricDMD.h for documentation.` The comments could be removed also from existing files like DMD.h and NonuniformDMD.h, since they are just duplicated. I wonder if there is a better design, perhaps with `getParametricDMD` being a method of the base `DMD` class, but we can leave that question for another PR.

Reply:
I'm not sure of the original intention of parametricDMD being its own class.  It would be my preference to leave the comments here, so that it is easy to find out what each one does, since this function provides essential functionality to the class.  If noone else agrees, I am happy to ditch them though.